_posts/2017-07-23-fastlane-travis-weekly-deploys.md

```diff
@@ -0,0 +1,314 @@
+
+The downside of this choice is that the process of uploading is not inside the main repo, and can go out of sync with the main app.
+
+## Setup
+
+I created a new repo, and added the [usual LICENSE and README][init], then started [working on a PR][pr] that added support for CI to run. Here are the general steps I needed to make work:
+
+* Downloading and setting up the application.
+* Ensuring signing will work.
+* Creating the build and shipping it to Testflight.
+* Notifications that it passed or succeeded.
+
+Finally I needed to document the process, which is what you're reading.
+
+## Downloading and setting up the application
+
+My initial thoughts were to use a submodule, but that option provides little advantage over cloning the repo itself so it's done inline. Our dependencies for the app live in Rubygems (Fastlane/CocoaPods), NPM (React Native) and CocoaPods (Artsy Mobile code), so I use the `before_install` and `before_script` section of the `.travis.yml` to set up our dependencies:
+
+```yml
+# Use a Mac build please
+language: objective-c
+osx_image: xcode8.2
+
+# Ensure that Fastlane is at the latest version
+before_install:
+- bundle update
+
+# Let fastlane set up the other dependency managers
+before_script:
+- bundle exec fastlane setup
+
+# Separate fastlane lanes so that they can be individually
+# tested one by one during development
+script:
+- bundle exec fastlane validate_env_vars
+- bundle exec fastlane stamp_plist
+- bundle exec fastlane ship
+```
+
+Note the `- bundle update`. As fastlane works against iTunes connect which is always changing locking the version doesn't make much sense. 
+
+## Ensuring signing will work.
+
+This one is a bit tricker, luckily I've already set up one of our apps to use [fastlane match][] and I can re-use that infrastructure. As it is a private repo, Travis did not have access to clone the repo. I fixed this by creating an access token for a user with read-only access to our match-codesigning repo, then exposed this as a private environment variable in CI which the Matchfile uses. E.g.
+
+```ruby
+git_url "https://#{ENV['GITHUB_SUBMODULES_USER']}@github.com/artsy/mobile_fastlane_match"
+
+# Instead of 
+# git_url "https://github.com/artsy/mobile_fastlane_match"
+```
+
+This is one of the highlights on fastlane's choice in building a DSL that which sits above a real programming language, you give users a lot of flexibility.
+
+Next up, I added a fastlane lane for code signing, and keychain setup. This just calls two setup functions.
+
+```ruby
+lane :setup_signing do
+  create_keychain(
+    name: ENV['MATCH_KEYCHAIN_NAME'],
+    password: ENV['MATCH_PASSWORD'],
+    default_keychain: true,
+    unlock: true,
+    timeout: 3600,
+    add_to_search_list: true
+  )
+
+  match(type: 'appstore',
+        keychain_name: ENV['MATCH_KEYCHAIN_NAME'],
```

Comment:
With the `setup_travis` action you can remove all options except for the `type` 👍